compiler/rustc_const_eval/src/interpret/terminator.rs

```diff
@@ -115,6 +88,45 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 }
             }
 
+            TailCall { ref func, ref args, fn_span: _ } => {
+                let old_frame_idx = self.frame_idx();
+
+                let (fn_val, args, fn_sig, fn_abi, with_caller_location) =
+                    self.prepare_fn_for_call(terminator, func, args)?;
+
+                // This is the "canonical" implementation of tails calls,
+                // a pop of the current stack frame, followed by a normal call
+                // which pushes a new stack frame, with the return address from
+                // the popped stack frame.
+                //
+                // Note that we can't use `pop_stack_frame` as it "executes"
+                // the goto to the return block, but we don't want to,
+                // only the tail called function should return to the current
+                // return block.
+                let Some(prev_frame) = self.stack_mut().pop()
+                    else { span_bug!(terminator.source_info.span, "empty stack while evaluating this tail call") };
+
+                let StackPopCleanup::Goto { ret, unwind } = prev_frame.return_to_block
+                    else { span_bug!(terminator.source_info.span, "tail call with the root stack frame") };
+
+                self.eval_fn_call(
+                    fn_val,
+                    (fn_sig.abi, fn_abi),
+                    &args,
+                    with_caller_location,
+                    &prev_frame.return_place,
+                    ret,
+                    unwind,
+                )?;
```

Comment:
Please add tests in `src/tools/miri/tests` to ensure tail calls work in Miri. You can run those tests with `./x.py test miri --stage 0`.

Reply:
Well, the test shows that tail calls in fact don't work in miri 😅 

Tail calls trip this assertion: https://github.com/rust-lang/rust/blob/8d361cbd91d827d703d2da46af8d89190ca1cb6a/src/tools/miri/src/concurrency/thread.rs#L186

I guess this is to be expected since we do not call `after_stack_pop` which maintains this cache...